concatenate for a nonzero axis operation

ianna opened the pull request:
address Issue https://github.com/scikit-hep/awkward-1.0/issues/184

ianna committed:
start on issue #184

ianna committed:
add a test

ianna committed:
merge RegularArrays in axis > 0

ianna committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into ianna/concatenate_for_nonzero_axis_operation

ianna committed:
remove debug cout

ianna committed:
remove merger

ianna committed:
remove merge functions from operations

ianna committed:
merge regular arrays at axis > 1

ianna committed:
separate merge at axis 0

ianna committed:
add other types to merge

ianna committed:
use data accessor for Index

ianna committed:
[skip ci] commit local development before rebase

ianna committed:
Merge branch 'master' into ianna/concatenate_for_nonzero_axis_operation

ianna committed:
merge with master done, comment out failing tests for now

ianna committed:
fix kernel tests, add ListOffsetArray merge tests in axis > 0

ianna committed:
merge more tests

ianna committed:
once again try to fix generatad kernel tests

ianna committed:
disable auto generated tests

ianna committed:
cleanup merged changes in numpyarray

ianna committed:
move axis upstream

ianna committed:
use mergemany where axis is 0, keep merge with an axis parameter

ianna commented on src/libawkward/array/ListArray.cpp:
```diff
@@ -963,6 +968,9 @@ namespace awkward {
       else {
+        // FIXME:???
+        return mergemany_as_union(others, 0, 0);
+
```
here is why:
```python
    def test_indexed_array_merge():
        one = awkward1.Array([[1, 2, 3], [None, 4], None, [None, 5]]).layout
        two = awkward1.Array([6, 7, 8]).layout
        three = awkward1.Array([[6.6], [7.7, 8.8]]).layout
    
>       assert awkward1.to_list(one.merge(two, 0)) == [[1, 2, 3], [None, 4], None, [None, 5], 6, 7, 8]
E       ValueError: cannot merge ListArray64 with NumpyArray
E       
E       (https://github.com/scikit-hep/awkward-1.0/blob/0.4.0/src/libawkward/array/ListArray.cpp#L976)

tests/test_0184-concatenate-operation.py:116: ValueError
```

jpivarski replied:
It's true that a ListArray64 can't be merged with a NumpyArray. The `merge` and `mergemany` functions are low-level and can only be used after `mergeable` certifies that the two arrays can be merged. If `mergeable` returns false, then `merge_as_union` must be used instead. The fact that the methods have these unwritten rules is why they're low-level: in Python, one should use `ak.concatenate`, and even most of our internal use of these functions are mediated through a `simplify` operation. I don't see any reason why this concatenate with `axis > 0` feature couldn't be implemented entirely by making a UnionArray and calling `simplify` on it.

ianna replied:
@jpivarski - thanks! done

ianna committed:
use ak.concatenate

ianna committed:
save before cleanup

ianna committed:
cleanup

ianna committed:
no need to pass axis to merge method

ianna committed:
revert changes where merge can be used without axis parameter

jpivarski commented on src/awkward1/operations/structure.py:
```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
+        raise ValueError(
+            "the 'axis' of concatenate must be non-negative"
+            + awkward1._util.exception_suffix(__file__)
+        )
```
Is this necessary? Your implementation uses `axis_wrap_if_negative`.

jpivarski replied:
Oh! Is it because you check

```python
if depth == axis
```

in `getfunction`? Maybe you could expose the `axis_wrap_if_negative` to Python, since this needs to be checked while descending, not when you've already reached the target axis. That could enable the feature for `axis < 0`.

jpivarski commented on src/awkward1/operations/structure.py:
```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
+                for x in arr[1:]:
+                    collapsed = batch[0].mergemany_as_union(batch[1:], 1)
+                    batch = [collapsed.merge_as_union(x, 1)]
+                out = batch[0].mergemany_as_union(batch[1:], 1)
```
I don't understand this: why is `mergemany_as_union` and `merge_as_union` being called once per array? Why shouldn't this implementation just be the following?

```python
out = inputs[0].mergemany_as_union(inputs[1:], 1)
```

ianna replied:
done

jpivarski replied:
I lost a huge comment here...

Looking at the original, I can see what it was doing:

```python
    batch = [contents[0]]
    for x in contents[1:]:
        if batch[-1].mergeable(x, mergebool=mergebool):
            batch.append(x)
        else:
            collapsed = batch[0].mergemany(batch[1:])
            batch = [collapsed.merge_as_union(x)]

    out = batch[0].mergemany(batch[1:])
```

It was accumulating batches of mutually `mergeable` arrays into a Python list (which does not scale with the size of the data) and then either calling `mergemany` on them all once at the end or once as soon as a non-mergeable array was encountered. Thus, the algorithm was _O(n)_ in the total length of the array: nothing was double-merged.

I haven't checked your implementation to see if it has the same characteristics. We'd want it to be able to batch mutually `mergeable` arrays and only `mergemany` them once (but at a deeper `axis`) and we'd want it to not attempt to `mergemany` on non-mergeable arrays. Keep in mind that the `mergemany` method raises RuntimeErrors (i.e. our internal errors) if you try to merge anything that can't be merged. If we don't compare explicitly verify that two arrays are `mergeable` before trying to merge them, it's a bug.

In the tests, can you verify that all combinations of the following arrays containing numbers and records (at `axis=1`) concatenate correctly?

```python
numbers = [
    awkward1.Array([[1.1, 2.2, 3.3], [], [4.4, 5.5]]),
    awkward1.Array([[10, 20], [30], [], [40, 50, 60]]),
    awkward1.Array(numpy.array([[101, 102, 103], [201, 202, 203]])),
]
records = [
    awkward1.Array([[{"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}], [], [{"x": 2.2, "y": [1, 2]}]]),
    awkward1.Array([[{"x": 3.3, "y": [1, 2, 3]}], [{"x": 4.4, "y": [2, 1]}, {"x": 5.5, "y": [1]}]]),
]
```

That is, lists like `numbers + records`, `numbers[:1] + records + numbers[1:]`, etc.? The two constraints that must be upheld are

   1. It must call `mergemany` exactly once on a contiguous run of same-type arrays (all numbers or all records).
   2. It must not call `mergemany` on different-type arrays.

You've also implemented a `mergemany_as_union`, which would be nice to call only once, rather than for every pair. Note that a UnionArray is `mergeable` with anything, since the merge would only make a UnionArray with more unique tags.

ianna replied:
done

jpivarski commented on src/awkward1/operations/structure.py:
```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
+                out = batch[0].mergemany_as_union(batch[1:], 1)
+                return lambda: (
+                    out,
+                )
```
Just

```python
return lambda: out
```

is fine. This looks like something that an auto-formatter like Black went nuts on.

ianna replied:
I think, it expects a tuple as an output:
```python
awkward1/operations/structure.py:778: in concatenate
    out = awkward1._util.broadcast_and_apply(arrays, getfunction, behavior)[0]
awkward1/_util.py:975: in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
...
>               assert isinstance(outcontent, tuple)
E               AssertionError

awkward1/_util.py:746: AssertionError

```

jpivarski replied:
You're right! It should be

```python
return lambda: (out,)
```

Python's tuple syntax can be subtle (because trailing commas are ignored in argument lists). But anyway, it doesn't need to be three lines.

ianna replied:
done

jpivarski commented on kernel-specification.yml:
```diff
@@ -2696,6 +2696,51 @@ kernels:
+              b = fromright[i + 1] - fromright[i]
+            else:
+              b = 0
+            tooffsets[i + 1] = tooffsets[i] + a + b
```
Thanks for filling in the normative definitions! Just a note, though: we use 4-space indent in Python. (I used to use 2-space indent in Python, too, since it's a syntactical element, it needs to be more visible and obvious. Also, Black and various style guides recommend 4 spaces. In C++, 2 spaces is fine because we always use curly brackets, even for one-line bodies.)

ianna replied:
done

jpivarski commented on kernel-specification.yml:
```diff
@@ -4689,6 +4734,54 @@ kernels:
+              else:
+                if ileft < leftlen:
+                  tag = 0
+                iright = iright + 1
```
Same thing: 4-space indent. (Your first indentation is 6 spaces.)

ianna replied:
done

jpivarski reviewed (approved):
This is a good suite of tests, particularly the fact that it agrees well with multidimensional NumPy arrays. So it appears to be fully working!

I wonder, though, about the `getfunction` implementation in Python. It seems to be calling `mergemany_as_union` and `merge_as_union` a lot more times than I'd think are necessary. Doesn't each `mergemany_as_union` merge all arrays in `inputs`? Why does that function need to be called for each array in `inputs`? This calculation ought to be _O(n)_ in the number of arrays _n_, but it looks like it's _O(n²)_.

There might also be low-hanging fruit in getting this to work for `axis < 0`; see inline comment.

ianna committed:
address Jim's comments

ianna committed:
revert to merge_as_union without axis

ianna committed:
revert other incidental changes

ianna committed:
Merge branch 'master' into ianna/concatenate_for_nonzero_axis_operation

ianna committed:
prepare to handle the negative axis

ianna committed:
use offsets_and_flaten

ianna committed:
merge all arrays at once, move their kernel code to C++ temporarily

ianna committed:
no need for pair, get length from Index itself

ianna committed:
really remove kernel code for merge

ianna committed:
st remains of kernel code removed

ianna committed:
clean up

ianna committed:
add numbers and records test

ianna commented on tests/test_0184-concatenate-operation.py:
```diff
@@ -0,0 +1,179 @@
+    numbers = [
+        awkward1.Array([[1.1, 2.2, 3.3], [], [4.4, 5.5], []]),
+        awkward1.Array([[10, 20], [30], [], [40, 50, 60]]),
+        awkward1.Array(numpy.array([[101, 102, 103], [201, 202, 203], [], []])),
```
@jpivarski - `broadcast_and_apply` doesn't like different lengths. What is a correct approach here?
```python
        numbers = [
            awkward1.Array([[1.1, 2.2, 3.3], [], [4.4, 5.5]]),
            awkward1.Array([[10, 20], [30], [], [40, 50, 60]]),
            awkward1.Array(numpy.array([[101, 102, 103], [201, 202, 203]])),
        ]
        records = [
            awkward1.Array([[{"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}], [], [{"x": 2.2, "y": [1, 2]}]]),
            awkward1.Array([[{"x": 3.3, "y": [1, 2, 3]}], [{"x": 4.4, "y": [2, 1]}, {"x": 5.5, "y": [1]}]]),
        ]
    
>       assert awkward1.to_list(awkward1.concatenate(numbers, 1)) == [

tests/test_0184-concatenate-operation.py:166: 

ValueError: cannot broadcast RegularArray of size 3 with RegularArray of size 4
E                           
E                           (https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/awkward1/_util.py#L754)
```

jpivarski replied:
So, `broadcast_and_apply` _should_ be requiring all input arrays to have equal lengths at all levels _before_ you apply your customization (through `getfunction`). So if `axis=10`, all of the input arrays would have to have the same lengths for 10 levels of recursion, and then any lengths they have deeper than that do not matter. For instance, with `axis=1`, all input arrays must have the same length (4 in this example). If `axis=2`, then they must all have the same outer array length as well as all nested list lengths, but then they can have whatever they want deeper than that.

These arrays can only do `axis=0` and `axis=1`. Looking closely, yes, I goofed up all their lengths. The first one has length 3, the second has length 4, and so you get the error message. Here's a better set:

```python
        numbers = [
            awkward1.Array([
                [1.1, 2.2, 3.3],
                [],
                [4.4, 5.5],
                [6.6, 7.7, 8.8, 9.9],
            ]),
            awkward1.Array([
                [10, 20],
                [30],
                [],
                [40, 50, 60],
            ]),
            awkward1.Array(numpy.array([
                [101, 102, 103],
                [201, 202, 203],
                [301, 302, 303],
                [401, 402, 403],
            ])),
        ]
        records = [
            awkward1.Array([
                [{"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
                [],
                [{"x": 2.2, "y": [1, 2]}],
                [{"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
            ]),
            awkward1.Array([
                [{"x": 0.0, "y": []}],
                [{"x": 1.1, "y": [1]}, {"x": 2.2, "y": [1, 2]}],
                [],
                [{"x": 3.3, "y": [1, 2, 3]}],
            ]),
        ]
```

They all have length 4. The lengths within each list differ wildly, but `broadcast_and_apply` should only do its length check for the first level when `axis=1`.

Here's what you should get from `ak.concatenate([numbers[0], numbers[1]], axis=1)`:

```python
[
    [1.1, 2.2, 3.3, 10, 20],
    [30],
    [4.4, 5.5],
    [6.6, 7.7, 8.8, 9.9, 40, 50, 60],
]
```

And here's what you should get from `ak.concatenate([numbers[0], records[0]], axis=1)`:

```python
[
    [1.1, 2.2, 3.3, {"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
    [],
    [4.4, 5.5, {"x": 2.2, "y": [1, 2]}],
    [6.6, 7.7, 8.8, 9.9, {"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
]
```

If you want an example that tests `broadcast_and_apply`'s length checking at more than one level, there's this:

```python
arrays = [
    ak.Array([[[0.0, 1.1, 2.2], []], [[3.3, 4.4]], [[5.5], [6.6, 7.7, 8.8, 9.9]]]),
    ak.Array([[[10, 20], [30]], [[40]], [[50, 60, 70], [80, 90]]]),
]
# nothing is required to have the same length
assert ak.concatenate(arrays, axis=0).tolist() == [
    [[0.0, 1.1, 2.2], []],
    [[3.3, 4.4]],
    [[5.5], [6.6, 7.7, 8.8, 9.9]],
    [[10, 20], [30]],
    [[40]],
    [[50, 60, 70], [80, 90]],
]
# the outermost arrays are required to have the same length, but nothing deeper than that
assert ak.concatenate(arrays, axis=1).tolist() == [
    [[0.0, 1.1, 2.2], [], [10, 20], [30]],
    [[3.3, 4.4], [40]],
    [[5.5], [6.6, 7.7, 8.8, 9.9], [50, 60, 70], [80, 90]],
]
# the outermost arrays and the first level are required to have the same length, but nothing deeper
assert ak.concatenate(arrays, axis=2).tolist() == [
    [[0.0, 1.1, 2.2, 10, 20], [30]],
    [[3.3, 4.4, 40]],
    [[5.5, 50, 60, 70], [6.6, 7.7, 8.8, 9.9, 80, 90]],
]
```

You should be able to concatenate `[arrays[0][:, :-1], arrays[1]]` at `axis=1`, but not at `axis=2`.

ianna replied:
ok, the tests pass, but there is a warning:
```python
tests/test_0184-concatenate-operation.py::test_list_offset_array_concatenate PASSED                                                                                                                                                                  [ 10%]
tests/test_0184-concatenate-operation.py::test_list_array_concatenate PASSED                                                                                                                                                                         [ 20%]
tests/test_0184-concatenate-operation.py::test_records_concatenate PASSED                                                                                                                                                                            [ 30%]
tests/test_0184-concatenate-operation.py::test_indexed_array_concatenate PASSED                                                                                                                                                                      [ 40%]
tests/test_0184-concatenate-operation.py::test_bytemasked_concatenate PASSED                                                                                                                                                                         [ 50%]
tests/test_0184-concatenate-operation.py::test_listoffsetarray_concatenate PASSED                                                                                                                                                                    [ 60%]
tests/test_0184-concatenate-operation.py::test_numpyarray_concatenate_axis0 PASSED                                                                                                                                                                   [ 70%]
tests/test_0184-concatenate-operation.py::test_numpyarray_concatenate PASSED                                                                                                                                                                         [ 80%]
tests/test_0184-concatenate-operation.py::test_numbers_and_records_concatenate PASSED                                                                                                                                                                [ 90%]
tests/test_0184-concatenate-operation.py::test_jims_test PASSED                                                                                                                                                                                      [100%]

===================================================================================================================== warnings summary =====================================================================================================================
tests/test_0184-concatenate-operation.py::test_numbers_and_records_concatenate
  /Users/yana/Projects/PR184-rebase/awkward-1.0/awkward1/_util.py:64: DeprecationWarning: In version 1.0.0 (target date: 2020-12-01), this will be an error.
  (Set ak.deprecations_as_errors = True to get a stack trace now.)
  
  ValueError: cannot broadcast: <class 'awkward1._ext.RecordArray'>, <class 'awkward1._ext.RecordArray'>
  
  (https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/awkward1/_util.py#L884)
    warnings.warn(message, DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============================================================================================================== 10 passed, 1 warning in 0.77s ===============================================================================================================

```

edit:
hmm... here is a future error. 

```python
>       assert ak.concatenate([numbers[0], records[0]], 1) == [
            [1.1, 2.2, 3.3, {"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
            [],
            [4.4, 5.5, {"x": 2.2, "y": [1, 2]}],
            [6.6, 7.7, 8.8, 9.9, {"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
        ]

tests/test_0184-concatenate-operation.py:193: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.pyenv/versions/3.8.1/lib/python3.8/site-packages/numpy/lib/mixins.py:25: in func
    return ufunc(self, other)
awkward1/highlevel.py:1369: in __array_ufunc__
    return awkward1._connect._numpy.array_ufunc(ufunc, method, inputs, kwargs)
awkward1/_connect/_numpy.py:173: in array_ufunc
    out = awkward1._util.broadcast_and_apply(
awkward1/_util.py:996: in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
awkward1/_util.py:759: in apply
    outcontent = apply(nextinputs, depth + 1)
awkward1/_util.py:847: in apply
    outcontent = apply(nextinputs, depth + 1)
awkward1/_util.py:659: in apply
    outcontents.append(apply(nextinputs, depth))
awkward1/_util.py:886: in apply
    deprecate(exception, "1.0.0", "2020-12-01")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exception = ValueError("cannot broadcast: <class 'awkward1._ext.RecordArray'>, <class 'awkward1._ext.RecordArray'>\n\n(https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/awkward1/_util.py#L884)"), version = '1.0.0', date = '2020-12-01'

    def deprecate(exception, version, date=None):
        if awkward1.deprecations_as_errors:
>           raise exception
E           ValueError: cannot broadcast: <class 'awkward1._ext.RecordArray'>, <class 'awkward1._ext.RecordArray'>
E           
E           (https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/awkward1/_util.py#L884)

awkward1/_util.py:54: ValueError

```

I've also removed all kernels I wrote for this PR to make sure this code was not causing the dereference error. The PR is pretty minimal now. It relies on the `offsets_and_flattened` and `broadcast_and_apply`. I can try to move this PR tests after the 0397 test, but apart from that I'm running out of the ideas on how to catch the dereference bug. Could it be that the garbage collector is simply delayed?

ianna replied:
fixed

ianna reviewed (commented):
@jpivarski - please, have a look at the example results. Is it what you'd expect?

ianna committed:
add Jim's tests

ianna committed:
comment out added tests to check if this will bring the dereference error back

ianna committed:
add the tests back

ianna committed:
cleanup tests

ianna committed:
fix deprication warning

ianna committed:
avoid code duplication by moving it to concatenate_here

ianna committed:
re-introduce kernel functions

sort kernel order

disable auto generated test for awkward_ListOffsetArray_merge_offsets:

% python dev/generate-tests.py
Generating Python kernels
Traceback (most recent call last):
  File "dev/generate-tests.py", line 560, in <module>
    specdict = readspec()
  File "dev/generate-tests.py", line 185, in readspec
    specdict[childfunc["name"]] = Specification(
  File "dev/generate-tests.py", line 38, in __init__
    self.tests = self.gettests(testdata)
  File "dev/generate-tests.py", line 142, in gettests
    funcPy(**temp)
  File "<string>", line 1286, in awkward_ListOffsetArray_merge_offsets
KeyError: 0

cleanup

ianna committed:
cleanup
